# Assignment 3

===

# I. Social Indicator Data 

In [1]:
# Importing packages
import pandas as pd
import seaborn as sns
import numpy as num
from scipy import stats as st

## 1. Download the data set “social_indicator.txt” in Canvas.

### a. Read it with python. Set the first column as the index when reading csv. (1 point for code)

In [2]:
# Reading in dataset and setting first column as the index
social = pd.read_csv('social_indicator.txt', index_col = 0, sep="\t")

### b.How many observations are there in the dataset? (1 point for answer)

There appears to be 155 observations in the social_indicator dataset.

### c.	Sort the row index into ascending order. (1 point for code)

In [3]:
# Sorting row index in ascending order(Default)
social.sort_values("Index", ascending = True, inplace = True)

### d.	Sort the column index lexicographically. (1 point for code)

In [4]:
# Sorting columns lexicographically
social.sort_index(axis=1, inplace = True)

### e.	Make a column ‘West’ with True and False values indicating whether the region is west. Then delete the original region column. (1 point for code)

In [5]:
# Defining the West column and creating it in the social dataset
def west (row):
    if row['region'] == 'West' :
      return True
    return False

social['West'] = social.apply(west, axis=1)

# Dropping the region column
social.drop(columns = 'region', inplace = True)

### f.	We are not interested in studying fertility rate (‘totalfertilityrate’) and contraception (‘contraception’). Drop these two columns. (1 point for code)

In [6]:
# Dropping the fertility rate and contraception columns
social.drop(columns = ['totalfertilityrate', 'contraception'], inplace = True)

### g.	Which variable is the most correlated with GDP per capita? Which variable is the least correlated with GDP per capita? (1 point for code; 2 points for answer)

In [7]:
# Finding all column's correlation with gdpPerCapita
corr_gdp = social.corrwith(social['gdpPerCapita'])

# Sorting correlation values
corr_gdp.sort_values(ascending = False)

gdpPerCapita              1.000000
lifeFemale                0.604029
lifeMale                  0.592267
economicActivityFemale    0.052964
economicActivityMale     -0.167231
illiteracyFemale         -0.457012
illiteracyMale           -0.471689
infantMortality          -0.584060
West                           NaN
dtype: float64

According to the chart above, it appears as if the 'lifeFemale' column is the most correlated with GDP per capita while the 'economicActivityFemale' column is the least correlated with GDP per capita (this of course is moving past that the fact that the 'West' column is just composed of True or False values meaning it cannot be correlated in this way).

### h.	Does the male economic activity have a positive or negative correlation with GDP per capita? Did you expect the positive/negative correlation? If not, why do you think you are observing that correlation? (no points, this is just for your understanding)

The male economic activity column actually has a negative correlation with GDP per capita. This statistic surprised me because you would think that more money spent (in general), would correlate to a higher GDP per capita. I think we see this because men are able to spend money in every country frequently, so perhaps this is true in both high and low GDP per capita countries.

## 2.

In [8]:
social['geographic_location'].unique()

array(['Western Asia', 'Southern Asia', 'Northern Africa',
       'North America', 'Oceania', 'Western Africa', 'Middle Africa',
       'South America', 'Southern Europe', 'Southern Africa',
       'Western Europe', 'Eastern Africa', 'South-eastern Asia',
       'Northern Europe', 'Eastern Europe', 'Eastern Asia',
       'Northern America', 'Central Asia'], dtype=object)

### a.	Use the column ‘geographic_location’ to create a new column called ‘continent’. Merge the values of the ‘geographic column appropriately to obtain 6 distinct values for the ‘continent’ column – Asia, Africa, North America, South America, Europe and Oceania. Drop the column ‘geographic_location’ (2 points for code).

In [9]:
# Creating the continent column
def area (row):
    if row['geographic_location'] == 'Western Asia' :
      return 'Asia'
    if row['geographic_location'] == 'South-eastern Asia' :
      return 'Asia'
    if row['geographic_location'] == 'Southern Asia' :
      return 'Asia'
    if row['geographic_location'] == 'Eastern Asia' :
      return 'Asia'
    if row['geographic_location'] == 'Central Asia' :
      return 'Asia'
    if row['geographic_location'] == 'Northern Africa' :
      return 'Africa'
    if row['geographic_location'] == 'Western Africa' :
      return 'Africa'
    if row['geographic_location'] == 'Middle Africa' :
      return 'Africa'
    if row['geographic_location'] == 'Southern Africa' :
      return 'Africa'
    if row['geographic_location'] == 'Eastern Africa' :
      return 'Africa'
    if row['geographic_location'] == 'North America' :
      return 'North America'
    if row['geographic_location'] == 'Northern America' :
      return 'North America'
    if row['geographic_location'] == 'South America' :
      return 'South America'
    if row['geographic_location'] == 'Northern Europe' :
      return 'Europe'
    if row['geographic_location'] == 'Southern Europe' :
      return 'Europe'
    if row['geographic_location'] == 'Eastern Europe' :
      return 'Europe'
    if row['geographic_location'] == 'Western Europe' :
      return 'Europe'
    return 'Oceania'

social['Continent'] = social.apply(area, axis=1)

# Dropping the geographic_location column
social.drop(columns = 'geographic_location', inplace = True)

### b.	Find the percentage of observations (i.e., rows) corresponding to each continent in the data (1 point for code, 1 point for answer).

In [10]:
# Compute the continent's percentage of observations
100 * social['Continent'].value_counts(normalize = True)

Asia             26.451613
Europe           22.580645
Africa           20.000000
North America    13.548387
Oceania           9.677419
South America     7.741935
Name: Continent, dtype: float64

According to the chart above, Asia comprises 26.45% of the observations, Africa comprises 20% of the observations, Europe comprises 18.06% of the observations, Oceania comprises 14.19% of the observations, North America comprises 13.54% of the observations,
and South America comprises 7.74% of the observations.

### c.	What is the rank of US amongst all countries in terms of GDP per capita? Which countries lie immediately above, and immediately below the US in the ranking in terms of GDP per capita? (2 points for code, 1 point for answer). 

In [11]:
# Finding the rank of the US according the gdpPerCapita
social['gdprank'] = social['gdpPerCapita'].rank(ascending = False)

# Found the top 15 countries
social.sort_values('gdprank').head(15)

country  economicActivityFemale  economicActivityMale  \
Index                                                                
43        Luxembourg                    37.3                  65.7   
100        Singapore                    51.5                  78.7   
50           Ireland                    41.1                  67.8   
22             Qatar                    27.5                  93.0   
116            Macau                    56.4                  81.1   
117      Switzerland                    56.8                  78.5   
139           Norway                    65.9                  76.8   
125    United States                    59.3                  74.9   
71            Brunei                    46.4                  82.2   
77         Hong Kong                    47.8                  76.0   
143          Denmark                    72.3                  83.2   
129      Netherlands                    60.2                  81.1   
149          Iceland                    76.8                  86.2   
76           Austria                    47.7                  69.5   
74           Germany                    47.3                  68.1   

       gdpPerCapita  illiteracyFemale  illiteracyMale  infantMortality  \
Index                                                                    
43           122740             1.000           1.000                6   
100          102742            13.700           4.100                5   
50            99239             1.000           1.000                6   
22            97262            20.100          20.800               17   
116           90606            14.616           5.412                8   
117           75880             1.000           1.000                5   
139           69171             0.000           0.000                5   
125           68309             2.232           2.244                7   
71            64405            16.600           7.400                9   
77            62839            11.800           4.000                5   
143           61478             1.000           1.000                7   
129           60461             1.000           1.000                6   
149           58151             1.000           1.000                5   
76            57891             1.000           1.000                6   
74            56956             1.000           1.000                6   

       lifeFemale  lifeMale   West      Continent  gdprank  
Index                                                       
43           79.7      73.1  False         Europe      1.0  
100          79.5      75.1  False           Asia      2.0  
50           79.4      74.0  False         Europe      3.0  
22           75.4      70.0  False           Asia      4.0  
116          80.1      75.1  False           Asia      5.0  
117          81.8      75.3  False         Europe      6.0  
139          80.6      74.8  False         Europe      7.0  
125          80.1      73.4  False  North America      8.0  
71           78.1      73.4  False           Asia      9.0  
77           81.8      76.1  False           Asia     10.0  
143          78.3      73.0  False         Europe     11.0  
129          80.6      75.0  False         Europe     12.0  
149          81.2      77.4  False         Europe     13.0  
76           80.1      73.7  False         Europe     14.0  
74           79.9      73.4  False         Europe     15.0

The US ranks 8th in the world for GDP per capita. In 7th place is Norway and in 9th place is Brunei.

### d.	Find all the countries that have a lower GDP per capita than the US, despite having lower illiteracy rates (for both genders), higher economic activity (for both genders), higher life expectancy (for both genders), and lower infant mortality rate than the US? (2 points for code, 1 point for answer)

In [12]:
social[(social.gdpPerCapita < 68309	) & (social.illiteracyFemale < 2.232) & (social.illiteracyMale < 2.244) & (social.economicActivityFemale > 59.3) & (social.economicActivityMale > 74.9) & (social.lifeFemale > 80.1) & (social.lifeMale > 73.4) & (social.infantMortality < 7)]

country  economicActivityFemale  economicActivityMale  \
Index                                                              
129    Netherlands                    60.2                  81.1   
147         Sweden                    75.6                  80.0   
149        Iceland                    76.8                  86.2   

       gdpPerCapita  illiteracyFemale  illiteracyMale  infantMortality  \
Index                                                                    
129           60461               1.0             1.0                6   
147           55566               1.0             1.0                5   
149           58151               1.0             1.0                5   

       lifeFemale  lifeMale   West Continent  gdprank  
Index                                                  
129          80.6      75.0  False    Europe     12.0  
147          80.8      76.2  False    Europe     16.0  
149          81.2      77.4  False    Europe     13.0

Sweden, Iceland, and Netherlands all have a lower GDP per capita despite ranking better than the US in some particular other categories.

### e.	Which country in North America has the second highest GDP per capita in the continent? What is its rank (in terms of GDP per capita) amongst all the countries? (2 points for code, 1 point for answer)

In [13]:
social[(social.Continent == 'North America')].sort_values('gdpPerCapita', ascending = False)

country  economicActivityFemale  economicActivityMale  \
Index                                                                      
125          United States                    59.3                  74.9   
121                 Canada                    57.6                  72.4   
140                Bahamas                    67.0                  81.2   
60                  Panama                    42.5                  81.3   
67     Trinidad and Tobago                    44.9                  75.5   
29              Costa Rica                    33.3                  79.2   
44                  Mexico                    37.4                  83.9   
61      Dominican Republic                    42.9                  82.6   
132               Barbados                    61.4                  73.4   
81             Saint Lucia                    48.5                  78.9   
111             Martinique                    54.1                  65.7   
38             Puerto Rico                    35.1                  64.0   
135                Jamaica                    62.4                  74.0   
104             Guadeloupe                    52.3                  66.9   
52                    Cuba                    41.2                  72.6   
49             El Salvador                    40.6                  79.8   
10               Guatemala                    18.4                  84.0   
33                  Belize                    34.0                  79.0   
37                Honduras                    34.7                  86.4   
30               Nicaragua                    33.4                  84.5   
86                   Haiti                    49.0                  80.3   

       gdpPerCapita  illiteracyFemale  illiteracyMale  infantMortality  \
Index                                                                    
125           68309             2.232           2.244                7   
121           51713            35.000           1.000                6   
140           34046             2.000           1.500               14   
60            30388             9.800           8.600               21   
67            25883             3.000           1.200               14   
29            20667             5.000           5.300               12   
44            20266            12.600           8.200               31   
61            19799            17.800          18.000               34   
132           14329             3.200           2.000                9   
81            13942             9.400          10.500               18   
111           12717             6.480           7.656                7   
38            12213             9.936           7.788                9   
135           10294            10.900          19.200               12   
104            9145             9.576           9.768                8   
52             9099             4.700           3.800                9   
49             8891            30.200          26.500               39   
10             8648            51.400          37.500               40   
33             6133            23.472          21.252               30   
37             5698            27.300          27.400               35   
30             5648            33.400          35.400               44   
86             2962            57.800          52.000               82   

       lifeFemale  lifeMale   West      Continent  gdprank  
Index                                                       
125          80.1      73.4  False  North America      8.0  
121          81.8      76.1  False  North America     20.0  
140          77.1      70.5  False  North America     43.0  
60           76.4      71.8  False  North America     49.0  
67           76.2      71.5  False  North America     55.0  
29           79.2      74.5  False  North America     63.0  
44           75.5      69.5  False  North America     65.0  
61           73.1 

Canada has the second highest gdpPerCapita in the North American continent. Canda's global rank for gdpPerCapita is 20.

## 3. Assume that all numeric columns, except GDP per capita, are social indicators.

### a.	Which country has the highest GDP per capita? Let us call it country G. (1 point for code, 1 point for answer)

In [14]:
social.sort_values('gdpPerCapita', ascending = False).head(10)

country  economicActivityFemale  economicActivityMale  \
Index                                                                
43        Luxembourg                    37.3                  65.7   
100        Singapore                    51.5                  78.7   
50           Ireland                    41.1                  67.8   
22             Qatar                    27.5                  93.0   
116            Macau                    56.4                  81.1   
117      Switzerland                    56.8                  78.5   
139           Norway                    65.9                  76.8   
125    United States                    59.3                  74.9   
71            Brunei                    46.4                  82.2   
77         Hong Kong                    47.8                  76.0   

       gdpPerCapita  illiteracyFemale  illiteracyMale  infantMortality  \
Index                                                                    
43           122740             1.000           1.000                6   
100          102742            13.700           4.100                5   
50            99239             1.000           1.000                6   
22            97262            20.100          20.800               17   
116           90606            14.616           5.412                8   
117           75880             1.000           1.000                5   
139           69171             0.000           0.000                5   
125           68309             2.232           2.244                7   
71            64405            16.600           7.400                9   
77            62839            11.800           4.000                5   

       lifeFemale  lifeMale   West      Continent  gdprank  
Index                                                       
43           79.7      73.1  False         Europe      1.0  
100          79.5      75.1  False           Asia      2.0  
50           79.4      74.0  False         Europe      3.0  
22           75.4      70.0  False           Asia      4.0  
116          80.1      75.1  False           Asia      5.0  
117          81.8      75.3  False         Europe      6.0  
139          80.6      74.8  False         Europe      7.0  
125          80.1      73.4  False  North America      8.0  
71           78.1      73.4  False           Asia      9.0  
77           81.8      76.1  False           Asia     10.0

Luxembourg has the highest GDP per capita, so country G is Luxembourg.

### b.	We need to find the African country that is the closest to country G with regard to social indicators. Perform the following steps:
#### i.	Standardize the social indicator columns appropriately so that all of them are on the same scale (remember to exclude GDP per capita from social indicators). (2 points for code)


In [15]:
# Creating standardizing function
f = lambda x: (x-x.min())/(x.max() - x.min())

# Applying f function
soc_ind = social.iloc[:,1:9].apply(f)

# Dropping the gdpPerCapita column
soc_ind.drop(columns = 'gdpPerCapita', inplace = True)
soc_ind

economicActivityFemale  economicActivityMale  illiteracyFemale  \
Index                                                                   
1                    0.000000              0.703349          0.765991   
2                    0.056370              0.660287          0.548458   
3                    0.059752              0.868421          0.936123   
4                    0.066516              0.602871          0.561674   
5                    0.090192              0.684211          0.448238   
...                       ...                   ...               ...   
151                  0.873732              0.901914          1.000000   
152                  0.933484              0.624402          0.148678   
153                  0.936866              0.856459          0.341410   
154                  0.953777              0.885167          0.844714   
155                  1.000000              0.930622          0.853524   

       illiteracyMale  infantMortality  lifeFemale  lifeMale  
Index                                                         
1            0.459660         0.463855    0.440639  0.516908  
2            0.404255         0.120482    0.783105  0.818841  
3            0.748936         0.909639    0.157534  0.217391  
4            0.370213         0.246988    0.712329  0.760870  
5            0.316312         0.216867    0.705479  0.785024  
...               ...              ...         ...       ...  
151          1.000000         0.566265    0.180365  0.219807  
152          0.276596         0.596386    0.372146  0.400966  
153          0.231206         0.120482    0.794521  0.811594  
154          0.600000         0.644578    0.212329  0.229469  
155          0.719149         0.668675    0.221461  0.229469  

[155 rows x 7 columns]

#### ii.	Compute the Manhattan distance between country G and each of the African countries. (2 points for code)

In [16]:
ind = social.index[social['country'] == 'Luxembourg']
africa = social[social['Continent'] == 'Africa']

africa_ind = africa.iloc[:,1:9].apply(f)
africa_ind.drop(columns = 'gdpPerCapita', inplace = True)

africaminusl = africa_ind - soc_ind.loc[ind[0],:]
africaminusl = africaminusl.pow(2)
distances = africaminusl.sum(axis=1)
distances_sorted = distances.sort_values()

#### iii.	Identify the African country, say country A, with the least Manhattan distance. (1 point for code, 1 point for answer)

In [17]:
distances_sorted.index[0]

social.loc[45]

country                   Mauritius
economicActivityFemale         38.8
economicActivityMale           80.1
gdpPerCapita                  22030
illiteracyFemale               21.2
illiteracyMale                 12.9
infantMortality                  15
lifeFemale                     75.0
lifeMale                       68.3
West                          False
Continent                    Africa
gdprank                        62.0
Name: 45, dtype: object

The African country with the least Manhattan distance to Luxembourg is Mauritius, so this is country A.

### c.	Does country A have the highest GDP per capita in Africa? (1 point for code, 1 point for answer)

In [18]:
africa.sort_values('gdpPerCapita', ascending = False)

country  economicActivityFemale  economicActivityMale  \
Index                                                                           
118                  Seychelles                    57.0                  78.8   
45                    Mauritius                    38.8                  80.1   
59            Equatorial Guinea                    42.4                  89.1   
55                     Botswana                    41.7                  75.4   
13                        Egypt                    21.6                  73.4   
93                 South Africa                    50.4                  79.6   
56                      Reunion                    41.9                  62.6   
4                       Algeria                     7.8                  76.4   
12                      Tunisia                    20.3                  75.4   
84                      Namibia                    49.0                  68.6   
17                      Morocco                    25.1                  74.9   
51                   Cape Verde                    41.1                  85.1   
57                  Ivory Coast                    42.0                  88.7   
42                      Nigeria                    37.2                  78.3   
88                        Congo                    49.2                  68.4   
19                        Sudan                    25.9                  74.6   
16                      Senegal                    24.3                  83.1   
122                       Benin                    57.8                  90.0   
23                    Swaziland                    27.7                  64.3   
124                    Ethiopia                    58.4                  84.7   
68                       Guinea                    44.9                  79.9   
73                     Zimbabwe                    46.7                  77.7   
66                       Gambia                    44.8                  70.2   
151                Burkina Faso                    79.4                  88.9   
94                         Togo                    50.8                  77.5   
148                Sierra Leone                    76.0                  80.0   
18                         Chad                    25.3                  91.2   
154                  Mozambique                    86.5                  88.2   
142    Central African Republic                    68.8                  87.0   
146                      Malawi                    75.3                  80.6   
155                     Burundi                    90.6                  90.1   

       gdpPerCapita  illiteracyFemale  illiteracyMale  infantMortality  \
Index                                                                    
118           26119            15.120          15.180               12   
45            22030            21.200          12.900               15   
59            18236            31.900          10.400              107   
55            18113            40.100          19.500               56   
13            13083            61.200          36.400               54   
93            12442            18.300          18.100               48   
56            11854            20.304          21.912                7   
4             11433            51.000          26.100               44   
12            10594            45.400          21.400               37   
84             9542            25.200          17.688               60   
17             8027            69.000          43.400               51   
51             6837            36.200          18.600               41   
57             5642            70.000          50.100               86   
42             5280            52.700          32.700               77   
88             4169            32.800          16.900               90   
19             4082            65.400          42.300               71   
16             362

No, country A (Mauritius) does not have the highest GDP per capita in Africa, Seychelles does.

### d.	Find the correlation between ‘Manhattan distance from country G’ and ‘GDP per capita’ for African countries. (1 point for code, 1 point for answer)

In [19]:
africa['manhattan'] = distances

africa.corrwith(africa['gdpPerCapita'])

<ipython-input-19-4ca98b094f3e>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  africa['manhattan'] = distances


economicActivityFemale   -0.323698
economicActivityMale     -0.204810
gdpPerCapita              1.000000
illiteracyFemale         -0.628112
illiteracyMale           -0.589864
infantMortality          -0.684109
lifeFemale                0.673534
lifeMale                  0.642327
West                           NaN
gdprank                  -0.980970
manhattan                -0.660251
dtype: float64

The correlation between the Mahattan distances and the GDP per capita of African countries is about -0.66. This means that the relationship is relatively strongly correlated.

### e.	Based on the correlation coefficient in (d), do you think African countries should try to emulate the social characteristics of country G? Justify your answer. (1 point for answer)

Yes! Because the correlation coefficient is this strongly negatively correlated, and these countries are the cloest to Luxembourg in terms of that correlated MAnhattan distance, the GDP per capita should improve if these African countries try to emulate the social charcteristcs of Luxembourg (country G).

## 4.	Suppose we are interested in studying life expectancy and want to find the countries with high life expectancy for both males and females. We define having a high life expectancy to be being in the top 25% for its corresponding gender. (the top 25th percentile included)

### a.	Make a subset of the data frame that only contains the countries with high life expectancy for both males and females. (2 points for code)

In [20]:
top25 = social[(social['lifeMale'] >= social['lifeMale'].quantile(0.75)) & (social['lifeFemale'] >= social['lifeFemale'].quantile(0.75))]

### b.	Which continents do not appear in the dataset obtained in (a)? Amongst the countries in the dataset obtained in (a), what percentage of them are European? (2 points for code, 1 point for answer)

In [21]:
top25['Continent'].unique()

array(['Europe', 'North America', 'Asia', 'Oceania'], dtype=object)

In [22]:
100 * top25['Continent'].value_counts(normalize = True)

Europe           51.428571
Asia             22.857143
North America    20.000000
Oceania           5.714286
Name: Continent, dtype: float64

South America and Africa are both the only continents that do not appear in this dataset. Furthermore, European countries comprise 51% of the countries in this dataset.

### c.	Name the country having the least GDP per capita in the dataset obtained in (a). (1 point for code, 1 point for answer)

In [23]:
top25.sort_values('gdpPerCapita', ascending = True)

country  economicActivityFemale  economicActivityMale  \
Index                                                                       
52                     Cuba                    41.2                  72.6   
38              Puerto Rico                    35.1                  64.0   
111              Martinique                    54.1                  65.7   
132                Barbados                    61.4                  73.4   
29               Costa Rica                    33.3                  79.2   
40                   Greece                    36.1                  64.7   
34                   Kuwait                    34.3                  80.1   
41                    Spain                    37.0                  63.1   
75                   Cyprus                    47.6                  78.0   
70                   Israel                    45.6                  62.1   
36                    Italy                    34.6                  61.6   
114             New Zealand                    56.2                  74.4   
90                    Japan                    50.0                  77.7   
21                    Malta                    27.4                  74.1   
106          United Kingdom                    53.5                  71.9   
78                   France                    48.4                  62.7   
121                  Canada                    57.6                  72.4   
112                 Finland                    55.4                  66.9   
96     Netherlands Antilles                    51.1                  69.8   
108               Australia                    53.8                  74.0   
147                  Sweden                    75.6                  80.0   
74                  Germany                    47.3                  68.1   
76                  Austria                    47.7                  69.5   
149                 Iceland                    76.8                  86.2   
129             Netherlands                    60.2                  81.1   
143                 Denmark                    72.3                  83.2   
77                Hong Kong                    47.8                  76.0   
71                   Brunei                    46.4                  82.2   
125           United States                    59.3                  74.9   
139                  Norway                    65.9                  76.8   
117             Switzerland                    56.8                  78.5   
116                   Macau                    56.4                  81.1   
50                  Ireland                    41.1                  67.8   
100               Singapore                    51.5                  78.7   
43               Luxembourg                    37.3                  65.7   

       gdpPerCapita  illiteracyFemale  illiteracyMale  infantMortality  \
Index                                                                    
52             9099             4.700           3.800                9   
38            12213             9.936           7.788                9   
111           12717             6.480           7.656                7   
132           14329             3.200           2.000                9   
29            20667             5.000           5.300               12   
40            30495             4.700           1.700                8   
34            41507            25.100          17.800               14   
41            41546             3.900           1.800                7   
75            41595             9.000           2.046                7   
70            42570             6.696           2.442                7   
36            43376             2.400           1.400                7   
114           44226             1.000           1.000                7   
90            44585             1.000           1.000                4   
21            45042            11.088          10.758       

Cuba has the smalles GDP per capita in this dataset.

### d.	What percentage of countries (in the whole dataset) have lower GDP per capita than the country obtained in (c)? (1 point for code, 1 point for answer)

In [24]:
100 * ((social[social['gdpPerCapita'] < 9099].count())/social.count())

country                   31.612903
economicActivityFemale    31.612903
economicActivityMale      31.612903
gdpPerCapita              31.612903
illiteracyFemale          31.612903
illiteracyMale            31.612903
infantMortality           31.612903
lifeFemale                31.612903
lifeMale                  31.612903
West                      31.612903
Continent                 31.612903
gdprank                   31.612903
dtype: float64

The percentage of countries that have a lower GDP per capita than Cuba is 31.6%.

### e.	Make another subset of the data frame that only contains the countries with high life expectancy for males but not for females. List the names of the countries. (1 point for code; 1 point for answer)

In [25]:
mixed = social[(social['lifeMale'] >= social['lifeMale'].quantile(0.75)) & (social['lifeFemale'] <= social['lifeFemale'].quantile(0.75))]

The countries that fit these constraints are UAE, Belize, Guam, and Jamaica.

## 5.	Use function application method (lambda function) to find i) interquartile range ii) Max-Min for all the numeric columns. (2 points for code)

In [26]:
# IQR Function
iqr = lambda x: (x.quantile(0.75)- x.quantile(0.25))

# Max - Min Function
mm = lambda x: (x.max() - x.min())

In [27]:
# Applying the IQR function
social.iloc[:,1:9].apply(iqr)

# Applying the Max - Min function
social.iloc[:,1:9].apply(mm)

economicActivityFemale        88.7
economicActivityMale          41.8
gdpPerCapita              121968.0
illiteracyFemale              90.8
illiteracyMale                70.5
infantMortality              166.0
lifeFemale                    43.8
lifeMale                      41.4
dtype: float64

### a.	Identify the gender with a larger range of economic activity. (1 point for answer)

### b.	Identify the gender with a smaller interquartile range of illiteracy. (1 point for answer)


a: Women have a larger range of economic activity (88.7 > 41.8).
    
b: Men have a smaller interquartile range of illiteracy (19.9 < 28.3).

# II. PATIENT DATA

## 1.	Download the data sets ‘observations.csv,’ ‘conditions.csv,’ and ‘patients.csv’ from Canvas. Suppose we are interested in studying patients with prediabetes condition. Do not drop or compute any missing values. In ‘condition.csv’, the patient IDs are stored in column ‘PATIENT’, and the medical conditions are stored in column ‘DESCRIPTION’. In ‘patient.csv’, the patient IDs are stored in column ‘Id’.

In [28]:
patient = pd.read_csv('patients.csv')
condition = pd.read_csv('conditions.csv')
observation = pd.read_csv('observations.csv')

patient

Id  Age  DEATHDATE          SSN  \
0  65b24dfa-d99d-74ae-bd55-d363e509cb7c   27        NaN  999-54-6959   
1  6d95ba48-a8be-b4c8-cdf6-f3d89142217b   35        NaN  999-60-2854   
2  46d05bdc-d001-63cc-693c-d7f29d531cb1    7        NaN  999-49-1943   
3  527c52ef-be00-25b5-823b-b87005bf0d5f   54        NaN  999-40-6727   
4  2599cde2-f690-8596-3cb5-905d701f6987   32        NaN  999-45-7333   
5  3583d823-d754-41b3-462f-3fb199c7b183   61        NaN  999-60-3795   
6  e9922ddd-8e85-9837-9ac8-885c34075450   70        NaN  999-91-6093   
7  5754cd1b-a29d-86d0-ecc7-358c06f0bb08   33        NaN  999-30-7711   
8  7a8a97f0-61b2-bd81-e556-ddea3d2f55f0   45        NaN  999-46-1942   
9  f3d4b3f5-5ca8-d376-c3ae-f61c93b1d130   55        NaN  999-92-5582   

     DRIVERS    PASSPORT PREFIX         FIRST        LAST  SUFFIX  ...  \
0  S99996424  X45500897X    Mr.        Mac103    Runte676     NaN  ...   
1  S99933959  X72103595X    Mr.   Jefferey580  Murazik203     NaN  ...   
2        NaN         NaN    NaN       Ronnie7     Smith67     NaN  ...   
3  S99951719   X1779287X   Mrs.    Valarie790   Hackett68     NaN  ...   
4  S99931616  X84630753X    Mr.  Filiberto722     West559     NaN  ...   
5  S99972612  X52173182X    Mr.       Lynn917  Kuvalis369     NaN  ...   
6  S99941771  X14585646X   Mrs.   Charlena802   Ernser583     NaN  ...   
7  S99972151  X47570332X   Mrs.     Sandie760  Sanford861     NaN  ...   
8  S99997709  X43705682X    Mr.       Huey641  Fritsch593     NaN  ...   
9  S99929701  X85339645X    Mr.      Bruno518   Lehner980     NaN  ...   

                      BIRTHPLACE                    ADDRESS        CITY  \
0    Plymouth  Massachusetts  US    960 Wiegand Burg Apt 94  Barnstable   
1      Boston  Massachusetts  US         600 Grimes Station    Brockton   
2      Sharon  Massachusetts  US           665 Roob Gardens      Boston   
3     Danvers  Massachusetts  US            695 Cassin Path  Somerville   
4   Littleton  Massachusetts  US        226 Skiles Junction      Ludlow   
5     Everett  Massachusetts  US  162 Leuschke Route Apt 82     Wareham   
6  Somerville  Massachusetts  US          1042 Roberts Wynd     Waltham   
7   Rochester  Massachusetts  US            454 Kuphal View     Needham   
8  Fall River  Massachusetts  US       629 Haag Park Apt 43     Peabody   
9      Boston  Massachusetts  US              699 O'Kon Rue    Falmouth   

           STATE             COUNTY     ZIP        LAT        LON  \
0  Massachusetts  Barnstable County  2647.0  41.673798 -70.272381   
1  Massachusetts    Plymouth County  2301.0  42.083140 -71.014794   
2  Massachusetts     Suffolk County  2124.0  42.317223 -71.035288   
3  Massachusetts   Middlesex County  2140.0  42.360324 -71.116043   
4  Massachusetts     Hampden County     NaN  42.170529 -72.439595   
5  Massachusetts    Plymouth County     NaN  41.723518 -70.708105   
6  Massachusetts   Middlesex County  2451.0  42.395781 -71.177870   
7  Massachusetts     Norfolk County     NaN  42.323663 -71.225636   
8  Massachusetts       Essex County  1940.0  42.511934 -71.003301   
9  Massachusetts  Barnstable County     NaN  41.555082 -70.646859   

  HEALTHCARE_EXPENSES HEALTHCARE_COVERAGE  
0        7.398092e+05           1061.9100  
1        8.748948e+05          14231.5875  
2        1.729582e+05           1979.8425  
3        3.689242e+05         894656.7430  
4        1.358564e+05         524142.3695  
5        1.697380e+06          67212.6675  
6        2.181451e+06         685662.5575  
7        1.271127e+06         275801.0080  
8        1.933860e+06          10709.0850  
9        1.411730e+06           1298.4825  

[10 rows x 25 columns]

### a.	Print the patient IDs of all the patients with prediabetes condition. (2 points for code)

In [29]:
prediabetes = condition[condition['DESCRIPTION'] == 'Prediabetes']
prediabetes['PATIENT']

40     e9922ddd-8e85-9837-9ac8-885c34075450
112    6d95ba48-a8be-b4c8-cdf6-f3d89142217b
170    2599cde2-f690-8596-3cb5-905d701f6987
182    527c52ef-be00-25b5-823b-b87005bf0d5f
223    f3d4b3f5-5ca8-d376-c3ae-f61c93b1d130
Name: PATIENT, dtype: object

### b.	Make a subset of the data with only prediabetes patients. How many prediabetes patients are there? (1 point for code, 1 point for answer)

In [30]:
prediabetes = condition[condition['DESCRIPTION'] == 'Prediabetes']

prediabetes.count()

START          5
STOP           0
PATIENT        5
ENCOUNTER      5
CODE           5
DESCRIPTION    5
dtype: int64

There are 5 prediabetes patients in this dataset.

### c.	Find the patient with highest healthcare expenses. Does this patient have prediabetes condition? (1 point for code, 1 point for answer)

In [31]:
patient.iloc[patient['HEALTHCARE_EXPENSES'].argmax()]

condition[condition['PATIENT'] == "e9922ddd-8e85-9837-9ac8-885c34075450"]

START        STOP                               PATIENT  \
25   1969-08-29         NaN  e9922ddd-8e85-9837-9ac8-885c34075450   
26   1969-08-29  1970-09-04  e9922ddd-8e85-9837-9ac8-885c34075450   
31   1976-09-10  1979-09-14  e9922ddd-8e85-9837-9ac8-885c34075450   
35   1979-05-10         NaN  e9922ddd-8e85-9837-9ac8-885c34075450   
36   1979-09-14  1985-09-20  e9922ddd-8e85-9837-9ac8-885c34075450   
40   1982-09-16         NaN  e9922ddd-8e85-9837-9ac8-885c34075450   
41   1982-09-16         NaN  e9922ddd-8e85-9837-9ac8-885c34075450   
43   1982-09-17         NaN  e9922ddd-8e85-9837-9ac8-885c34075450   
47   1987-07-03         NaN  e9922ddd-8e85-9837-9ac8-885c34075450   
59   1990-05-25  1997-07-25  e9922ddd-8e85-9837-9ac8-885c34075450   
66   1999-07-30  2001-07-06  e9922ddd-8e85-9837-9ac8-885c34075450   
67   1999-07-30  2001-07-06  e9922ddd-8e85-9837-9ac8-885c34075450   
71   2001-07-06  2002-05-10  e9922ddd-8e85-9837-9ac8-885c34075450   
76   2004-07-23  2005-07-29  e9922ddd-8e85-9837-9ac8-885c34075450   
77   2004-07-23  2020-10-23  e9922ddd-8e85-9837-9ac8-885c34075450   
78   2004-07-23  2006-08-04  e9922ddd-8e85-9837-9ac8-885c34075450   
79   2005-07-29  2006-08-04  e9922ddd-8e85-9837-9ac8-885c34075450   
86   2007-08-10  2011-09-02  e9922ddd-8e85-9837-9ac8-885c34075450   
90   2008-10-11         NaN  e9922ddd-8e85-9837-9ac8-885c34075450   
96   2009-08-21  2010-08-27  e9922ddd-8e85-9837-9ac8-885c34075450   
103  2010-08-27  2011-09-02  e9922ddd-8e85-9837-9ac8-885c34075450   
106  2011-09-02  2012-09-07  e9922ddd-8e85-9837-9ac8-885c34075450   
107  2011-09-02  2012-09-07  e9922ddd-8e85-9837-9ac8-885c34075450   
117  2012-09-07  2013-09-13  e9922ddd-8e85-9837-9ac8-885c34075450   
122  2013-09-13  2014-09-19  e9922ddd-8e85-9837-9ac8-885c34075450   
126  2013-12-10  2013-12-18  e9922ddd-8e85-9837-9ac8-885c34075450   
129  2014-09-19  2015-09-25  e9922ddd-8e85-9837-9ac8-885c34075450   
130  2014-09-19  2017-08-18  e9922ddd-8e85-9837-9ac8-885c34075450   
135  2015-09-25  2016-09-30  e9922ddd-8e85-9837-9ac8-885c34075450   
136  2015-09-25  2016-09-30  e9922ddd-8e85-9837-9ac8-885c34075450   
137  2016-05-15  2016-06-03  e9922ddd-8e85-9837-9ac8-885c34075450   
145  2016-09-30  2017-08-18  e9922ddd-8e85-9837-9ac8-885c34075450   
149  2017-08-08  2017-08-26  e9922ddd-8e85-9837-9ac8-885c34075450   
150  2017-08-18  2017-10-06  e9922ddd-8e85-9837-9ac8-885c34075450   
151  2017-10-06  2018-10-12  e9922ddd-8e85-9837-9ac8-885c34075450   
152  2017-10-06  2019-05-10  e9922ddd-8e85-9837-9ac8-885c34075450   
156  2018-01-28  2018-02-10  e9922ddd-8e85-9837-9ac8-885c34075450   
158  2018-10-12  2019-05-10  e9922ddd-8e85-9837-9ac8-885c34075450   
162  2019-05-01  2019-05-21  e9922ddd-8e85-9837-9ac8-885c34075450   
164  2019-05-10  2019-10-18  e9922ddd-8e85-9837-9ac8-885c34075450   
165  2019-06-05         NaN  e9922ddd-8e85-9837-9ac8-885c34075450   
181  2019-10-18  2020-10-23  e9922ddd-8e85-9837-9ac8-885c34075450   
187  2020-05-18  2020-06-11  e9922ddd-8e85-9837-9ac8-885c34075450   
192  2020-10-23         NaN  e9922ddd-8e85-9837-9ac8-885c34075450   
193  2020-10-23         NaN  e9922ddd-8e85-9837-9ac8-885c34075450   

                                ENCOUNTER       CODE  \
25   9f393518-88a0-9d88-c570-e2d06c81a392  224299000   
26   9f393518-88a0-9d88-c570-e2d06c81a392  160904001   
31   17c6852b-7aa2-021d-e88c-287573176283  160904001   
35   a4beb9fa-3fc1-e378-3498-a97992e3527a   19169002   
36   fcd4e8e1-3002-5b06-3b5d-46b41656c179   73595000   
40   b3044fcf-42bd-13e1-eb06-c385976c01fc   15777000   
41   b3044fcf-42bd-13e1-eb06-c385976c01fc  271737000   
43   25058a4a-823f-9688-6b2d-a89b64b760f3  266948004   
47   9a58cb57-5998-56b6-95cd-badfe3ac5eb3   95417003   
59   8e8099dc-f013-006d-afcd-d46cc5cd381a   73595000   
66   8c2b1057-ec1a-2e24-68f3-22f095f632f6  160904001   
67   8c2b1057-ec1a-2e24-68f3-22f095f632f6   73595000   
71   2cdefde5-006d-1901-607b-189bb59969cd  160904001   
76   dc009ab7-4517-9163-3bd3-545478b0140e  16090400

Yes! This patient has current prediabetes diagnosis from 1982.

## 2.	Suppose we are interested in studying the age of prediabetes patients. In ‘patient.csv’, the birth dates are stored in column ‘BIRTHDATE’. 

### a.	What is the mean age of patients without prediabetes condition? (1 point for code, 1 point for answer)

In [32]:
nonpred = patient.query('Id not in {0}'.format(prediabetes['PATIENT'].to_list()))

nonpred.mean()

Age                    3.460000e+01
DEATHDATE                       NaN
SUFFIX                          NaN
ZIP                    2.237000e+03
LAT                    4.211003e+01
LON                   -7.084894e+01
HEALTHCARE_EXPENSES    1.163027e+06
HEALTHCARE_COVERAGE    7.135290e+04
dtype: float64

The mean age of patients without a prediabetes condition is 34.6.

### b.	What is the mean age of patients with prediabetes condition? (1 point for code, 1 point for answer)



In [33]:
pred = patient.query('Id in {0}'.format(prediabetes['PATIENT'].to_list()))

pred.mean()

Age                        49.200000
DEATHDATE                        NaN
SUFFIX                           NaN
ZIP                      2297.333333
LAT                        42.112971
LON                       -71.279032
HEALTHCARE_EXPENSES    994571.362200
HEALTHCARE_COVERAGE    423998.348000
dtype: float64

The mean age of patients with the prediabetes condition is 49.2.

### c.	Is the mean age of patients without prediabetes condition significantly lower than the mean age of patients with prediabetes condition? Justify your answer with a one-sided T-Test. Assume our significance level is 10%. (1 point for code, 1 point for answer)

In [34]:
st.ttest_ind(nonpred['Age'], pred['Age'], alternative = 'greater')

Ttest_indResult(statistic=-1.277071055979225, pvalue=0.88130322596058)

Because the p-value is 0.88, which is below 0.10 (as our significance level), the mean age of patients without prediabetes condion IS significantly lower than the mean age of patients with prediabetes condtion.

## 3.	Suppose we are interested in studying the BMI of prediabetes patients. In ‘observations.csv’, there are different observation types stored in column ‘OBSERVATION’, with the corresponding observations stored in column ‘VALUE’. Patient IDs are stored in column ‘PATIENT’. 

### a.	Sort only the values corresponding to BMI observations. A BMI observation is any row with ‘DESCRIPTION’ equal to ‘Body Mass Index’. (2 points for code)



In [39]:
bmi = observation.query("DESCRIPTION in 'Body Mass Index'").sort_values('VALUE')
bmi

DATE                               PATIENT  \
1153  2019-08-24T19:12:27Z  46d05bdc-d001-63cc-693c-d7f29d531cb1   
1165  2020-08-29T19:12:27Z  46d05bdc-d001-63cc-693c-d7f29d531cb1   
1111  2018-08-18T19:12:27Z  46d05bdc-d001-63cc-693c-d7f29d531cb1   
1187  2021-09-04T19:12:27Z  46d05bdc-d001-63cc-693c-d7f29d531cb1   
975   2018-02-17T19:12:27Z  46d05bdc-d001-63cc-693c-d7f29d531cb1   
...                    ...                                   ...   
1522  2016-12-04T07:49:15Z  527c52ef-be00-25b5-823b-b87005bf0d5f   
1659  2017-12-10T07:49:15Z  527c52ef-be00-25b5-823b-b87005bf0d5f   
1935  2020-12-27T07:49:15Z  527c52ef-be00-25b5-823b-b87005bf0d5f   
1796  2019-12-22T07:49:15Z  527c52ef-be00-25b5-823b-b87005bf0d5f   
3202  2021-05-05T14:02:41Z  f3d4b3f5-5ca8-d376-c3ae-f61c93b1d130   

                                 ENCOUNTER     CATEGORY     CODE  \
1153  0fa28725-c1e7-8034-2d6f-d7aea54fa52b  vital-signs  39156-5   
1165  6f408b43-f759-6391-884b-f73083b05cd7  vital-signs  39156-5   
1111  4377eddc-1913-63d7-a866-0b15bb036c04  vital-signs  39156-5   
1187  9c308da0-83ee-c854-56ba-eddb3146bf00  vital-signs  39156-5   
975   d760e6a7-820d-bb96-c03a-88867fd1944d  vital-signs  39156-5   
...                                    ...          ...      ...   
1522  0d55f589-83f4-03ef-cb99-5760bee94e01  vital-signs  39156-5   
1659  7cc00ee1-ca71-a2f3-5bb0-701ad8885515  vital-signs  39156-5   
1935  f92b2e5e-0a1b-eb32-64a2-ad1b924d88d8  vital-signs  39156-5   
1796  16419c7b-5484-c9b1-42b0-f8e5c6d1622a  vital-signs  39156-5   
3202  002926b0-ff29-aa4f-ca55-f4ccaa5a26e8  vital-signs  39156-5   

          DESCRIPTION VALUE  UNITS     TYPE  
1153  Body Mass Index  16.0  kg/m2  numeric  
1165  Body Mass Index  16.2  kg/m2  numeric  
1111  Body Mass Index  16.8  kg/m2  numeric  
1187  Body Mass Index  17.1  kg/m2  numeric  
975   Body Mass Index  17.4  kg/m2  numeric  
...               ...   ...    ...      ...  
1522  Body Mass Index  30.6  kg/m2  numeric  
1659  Body Mass Index  30.6  kg/m2  numeric  
1935  Body Mass Index  30.6  kg/m2  numeric  
1796  Body Mass Index  30.6  kg/m2  numeric  
3202  Body Mass Index  30.6  kg/m2  numeric  

[72 rows x 9 columns]

### b.	What is the name of the patient with highest BMI? Note that patients may have more than one BMI recorded, and we only consider the patient with the highest occurrence in this question. (2 points for code, 1 point for answer)



In [46]:
bmi['VALUE'] =  bmi['VALUE'].astype(float)

bmi.iloc[bmi['VALUE'].argmax()]

patient[patient['Id'] == "f3d4b3f5-5ca8-d376-c3ae-f61c93b1d130"]

Id  Age  DEATHDATE          SSN  \
9  f3d4b3f5-5ca8-d376-c3ae-f61c93b1d130   55        NaN  999-92-5582   

     DRIVERS    PASSPORT PREFIX     FIRST       LAST  SUFFIX  ...  \
9  S99929701  X85339645X    Mr.  Bruno518  Lehner980     NaN  ...   

                  BIRTHPLACE        ADDRESS      CITY          STATE  \
9  Boston  Massachusetts  US  699 O'Kon Rue  Falmouth  Massachusetts   

              COUNTY ZIP        LAT        LON HEALTHCARE_EXPENSES  \
9  Barnstable County NaN  41.555082 -70.646859         1411729.958   

  HEALTHCARE_COVERAGE  
9           1298.4825  

[1 rows x 25 columns]

The name of the patient with the highest BMI is Mr. Bruno Lehner (or Bruno518 Lehner980).

### c.	Does the patient have prediabetes condition? (1 point for answer)

In [48]:
condition.loc[condition['PATIENT'] == "f3d4b3f5-5ca8-d376-c3ae-f61c93b1d130"]

START        STOP                               PATIENT  \
219  1984-10-10         NaN  f3d4b3f5-5ca8-d376-c3ae-f61c93b1d130   
220  1984-10-10         NaN  f3d4b3f5-5ca8-d376-c3ae-f61c93b1d130   
221  1990-11-14  2013-03-20  f3d4b3f5-5ca8-d376-c3ae-f61c93b1d130   
222  1994-12-07  1995-12-13  f3d4b3f5-5ca8-d376-c3ae-f61c93b1d130   
223  2003-01-22         NaN  f3d4b3f5-5ca8-d376-c3ae-f61c93b1d130   
224  2003-01-22  2004-01-28  f3d4b3f5-5ca8-d376-c3ae-f61c93b1d130   
225  2004-01-28         NaN  f3d4b3f5-5ca8-d376-c3ae-f61c93b1d130   
226  2004-01-28  2005-02-02  f3d4b3f5-5ca8-d376-c3ae-f61c93b1d130   
227  2005-02-02         NaN  f3d4b3f5-5ca8-d376-c3ae-f61c93b1d130   
228  2006-02-08  2007-02-14  f3d4b3f5-5ca8-d376-c3ae-f61c93b1d130   
229  2009-02-25  2010-03-03  f3d4b3f5-5ca8-d376-c3ae-f61c93b1d130   
230  2011-03-09  2012-03-14  f3d4b3f5-5ca8-d376-c3ae-f61c93b1d130   
231  2012-03-14  2013-03-20  f3d4b3f5-5ca8-d376-c3ae-f61c93b1d130   
232  2013-03-20  2014-03-26  f3d4b3f5-5ca8-d376-c3ae-f61c93b1d130   
233  2013-03-20  2014-03-26  f3d4b3f5-5ca8-d376-c3ae-f61c93b1d130   
234  2014-03-26  2015-04-01  f3d4b3f5-5ca8-d376-c3ae-f61c93b1d130   
235  2014-03-26  2021-05-05  f3d4b3f5-5ca8-d376-c3ae-f61c93b1d130   
236  2015-04-01  2016-04-06  f3d4b3f5-5ca8-d376-c3ae-f61c93b1d130   
237  2015-04-01  2017-04-12  f3d4b3f5-5ca8-d376-c3ae-f61c93b1d130   
238  2015-11-16  2015-11-27  f3d4b3f5-5ca8-d376-c3ae-f61c93b1d130   
239  2016-04-06  2017-04-12  f3d4b3f5-5ca8-d376-c3ae-f61c93b1d130   
240  2017-04-12  2018-04-18  f3d4b3f5-5ca8-d376-c3ae-f61c93b1d130   
241  2018-04-18  2019-04-24  f3d4b3f5-5ca8-d376-c3ae-f61c93b1d130   
242  2018-04-18  2020-04-29  f3d4b3f5-5ca8-d376-c3ae-f61c93b1d130   
243  2018-04-18  2019-04-24  f3d4b3f5-5ca8-d376-c3ae-f61c93b1d130   
244  2019-04-24  2020-04-29  f3d4b3f5-5ca8-d376-c3ae-f61c93b1d130   
245  2020-04-29  2021-05-05  f3d4b3f5-5ca8-d376-c3ae-f61c93b1d130   
246  2021-05-05         NaN  f3d4b3f5-5ca8-d376-c3ae-f61c93b1d130   

                                ENCOUNTER       CODE  \
219  4993ec63-32b2-a323-dee2-de2971080b8a   59621000   
220  4993ec63-32b2-a323-dee2-de2971080b8a  224295006   
221  7c2063ed-358e-257e-93b7-f2fa0c0213c4  422650009   
222  35fd73ef-fae9-263d-2873-f3d800985344  160904001   
223  ccfad290-2e11-f094-c025-9ae55888fd47   15777000   
224  ccfad290-2e11-f094-c025-9ae55888fd47  160904001   
225  b300fa89-267c-2de4-ec7b-398d49f78471  271737000   
226  66e8c5c4-3c29-40ea-2c47-34503d5cb5f2  160904001   
227  65a85584-f95e-0519-8fa2-b3254e8d8227   91302008   
228  00b0a667-6db0-7c98-7bfb-eb1c86417103  160904001   
229  9c29acab-8945-b73d-89fa-03ef750bc324  160904001   
230  84afe64f-a54c-cc3f-f712-0cf9f2a0f949  160903007   
231  64f32af2-1c7b-a548-6094-c88595d3745d  160903007   
232  62ecfcf6-b25d-13e0-7150-9712282f0be2  160903007   
233  62ecfcf6-b25d-13e0-7150-9712282f0be2   73595000   
234  d6b29c02-3aa3-f450-7d50-5ebde80ea98e  160903007   
235  d6b29c02-3aa3-f450-7d50-5ebde80ea98e  423315002   
236  36fe9184-887a-8436-5c18-55a6ffde4f07  160903007   
237  36fe9184-887a-8436-5c18-55a6ffde4f07   73595000   
238  c7b807b8-d175-86fe-18f9-0903d7c2fe71   10509002   
239  29528f46-b7ba-d29d-36e1-874fe664569d  160903007   
240  b2232e04-3429-2fe0-2db2-b50a2d4dcea1  160903007   
241  7f3d39df-e549-85df-7d92-a21c6a7ede56  160903007   
242  7f3d39df-e549-85df-7d92-a21c6a7ede56   73595000   
243  7f3d39df-e549-85df-7d92-a21c6a7ede56  706893006   
244  4b8ce9e3-52b5-1e7a-6be8-f9023b71cd69  160903007   
245  e8ed5e43-d652-0850-5605-a093ed11224b  160903007   
246  002926b0-ff29-aa4f-ca55-f4ccaa5a26e8  160904001   

                                          DESCRIPTION  
219                                      Hypertension  
220  Only received primary school education (finding)  
221                        Social isolation (finding)  
222                    Part-time employment (finding)  
223                                       Prediabetes  
224                    Part-time employment (fi

Yes! This patient still has a prediabetes condition that was never resolved.

### d.	Find the summary statistics of prediabetes patients’ BMIs. What is the standard deviation of BMIs. (1 point for code, 1 point for answer))



In [49]:
bmi.describe()

VALUE
count  72.000000
mean   27.450000
std     3.774675
min    16.000000
25%    26.900000
50%    28.550000
75%    30.125000
max    30.600000

The standard deviation is 3.774675.